### PR TITLE
Check digit

### LINKED ISSUE #1168: Check digit API
~~### _Feel free to comment with changes here. This is just a suggestion to get us started._~~

### Based on conversations that endpoints and responses have been updated ... but still feel free to comment. 😄 

We'll need 2 endpoints for the check digit API. Both should return `json` response and file uploads should allow for both `json` and `CSV` responses.

A few quick notes:

- the validation for a ULI can be found on page 20 of the [2018 FIG](https://www.consumerfinance.gov/data-research/hmda/static/for-filers/2018/2018-HMDA-FIG.pdf)
  - alphanumeric
  - max width of 45 characters
- ULI check digit is calculated using the ISO/IEC 7064, MOD 97-10 as it appears on the International Standard ISO/IEC 7064:2003, which is published by the International Organization for Standardization (ISO)
  - https://www.iso.org/obp/ui/#iso:std:iso-iec:7064:ed-1:v1:en
  - https://github.com/stvkoch/ISO7064-Mod-97-10

---

#### `/checkDigit` and `/checkDigit/csv`

`POST` - Accepts a single loan id input __or__ a file upload and responds with the check digit(s).

#### Single input

Example `POST` data:

```json
{
  "loanId": "1A2B3C4D5E6F"
}
```

Example single input response, in `json` format:

```json
{
  "loanId": "1A2B3C4D5E6F",
  "checkDigit": "22",
  "uli": "1A2B3C4D5E6F22"
}
```

#### File upload

Example file upload (`loans.txt`):
```
1A2B3C4D5E6F
2B3C4D5E6F7G
3C4D5E6F7G8H
4D5E6F7G8H9I
...
```

Example file upload response, in `json` format:
```json
{
  "loanIds": [
    {
      "loanId": "1A2B3C4D5E6F",
      "checkDigit": "22",
      "uli": "1A2B3C4D5E6F22"
    },
    {
      "loanId": "2B3C4D5E6F7G",
      "checkDigit": "33",
      "uli": "2B3C4D5E6F7G33"
    },
    {
      "loanId": "3C4D5E6F7G8H",
      "checkDigit": "44",
      "uli": "3C4D5E6F7G8H44"
    },
    {
      "loanId": "4D5E6F7G8H9I",
      "checkDigit": "55",
      "uli": "4D5E6F7G8H9I55"
    }
  ]
}
```

Example file upload response, in `CSV` format:
```csv
loanId, checkDigit, uli
1A2B3C4D5E6F, 22, 1A2B3C4D5E6F22
2B3C4D5E6F7G, 33, 2B3C4D5E6F7G33
3C4D5E6F7G8H, 44, 3C4D5E6F7G8H44
4D5E6F7G8H9I, 55, 4D5E6F7G8H9I55
```

---

### `/checkDigit/validate`

`POST` - Accepts a single ULI input __or__ a file upload and responds with true/false.

#### Single input

Example `POST` data:

```json
{
  "uli": "1A2B3C4D5E6F22"
}
```

Example single input response, in `json` format:
```json
{
  "isValid": true
}
```

#### File upload

Example file (`ulis.txt`):
```
1A2B3C4D5E6F22
2B3C4D5E6F7G33
3C4D5E6F7G8H44
4D5E6F7G8H9I55
...
```

Example file upload response, in `json` format:
```json
{
  "ulis": [
    {
      "uli": "1A2B3C4D5E6F22",
      "isValid": true
    },
    {
      "uli": "2B3C4D5E6F7G33",
      "isValid": true
    },
    {
      "uli": "3C4D5E6F7G8H44",
      "isValid": true
    },
    {
      "uli": "4D5E6F7G8H9I55",
      "isValid": false
    }
  ]
}
```

Example file upload response, in `CSV` format:
```csv
uli, isValid
1A2B3C4D5E6F22, true
2B3C4D5E6F7G33, true
3C4D5E6F7G8H44, true
4D5E6F7G8H9I55, false
```

### PR DESCRIPTION
**Pull request by jmarin**
Closes #1168

### CODE REVIEW

File: validation/src/main/scala/hmda/validation/engine/lar/ULI.scala [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    "x" -> 33,
+    "y" -> 34,
+    "z" -> 35
+  )
+
+  def convert(loanId: String): String = loanId
+    .map(_.toLower)
+    .map { c =>
+      if (!c.isDigit)
+        conversionTable(c.toString).toString
```
Comment (nickgrippin) [inline validation/src/main/scala/hmda/validation/engine/lar/ULI.scala]: Would it be better to do a conversion based on the ASCII value?  Something like `c.hashCode - 87`.  Though, a downside is that it's not immediately obvious that it follows the regulation.

Comment (jmarin) [reply validation/src/main/scala/hmda/validation/engine/lar/ULI.scala]: Yeah, that's a good suggestion, but I think it obscures the intent a bit. I'm ok with making this change if you think it's worth it.

Comment (nickgrippin) [reply validation/src/main/scala/hmda/validation/engine/lar/ULI.scala]: Definitely not a necessity, I'm ok with leaving it as-is for the transparency.

**Comment by nickgrippin** [comment]
LGTM, smoke tested all endpoints and the ULI follows regulations.

**Pull request merged**